drivers/ps2/ps2_npcx_controller.c

```diff
@@ -0,0 +1,354 @@
+
+	/* Write the data to be transmitted */
+	inst->PSDAT = value;
+	/* Apply the Request-to-send */
+	inst->PSOSIG &= ~BIT(NPCX_PSOSIG_WDAT(channel_id));
+	inst->PSOSIG |= BIT(NPCX_PSOSIG_CLK(channel_id));
+	if (k_sem_take(&data->tx_sync_sem, PS2_TRANSACTION_TIMEOUT) != 0) {
+		irq_disable(DT_INST_IRQN(0));
+		LOG_ERR("PS/2 Tx timeout");
+		/* Reset the shift mechanism */
+		inst->PSOSIG = NPCX_PS2_SHIFT_MECH_RESET;
+		/* Change the PS/2 module to receive mode */
+		inst->PSCON &= ~BIT(NPCX_PSCON_XMT);
+		/*
+		 * Restore the channel back to enable according to
+		 * channel_enabled_mask.
+		 */
+		inst->PSOSIG |= data->channel_enabled_mask;
+		irq_enable(DT_INST_IRQN(0));
+		goto timeout_invalid;
+	}
+
+	k_sem_give(&data->lock);
+	return 0;
+
+timeout_invalid:
+	k_sem_give(&data->lock);
+	return -ETIMEDOUT;
+}
+
+static int ps2_npcx_ctrl_is_rx_error(const struct device *dev)
+{
+	struct ps2_reg *const inst = HAL_PS2_INSTANCE(dev);
+	uint8_t status;
+
+	status = inst->PSTAT & (BIT(NPCX_PSTAT_PERR) | BIT(NPCX_PSTAT_RFERR));
+	if (status) {
+		if (status & BIT(NPCX_PSTAT_PERR))
+			LOG_ERR("RX parity error");
+		if (status & BIT(NPCX_PSTAT_RFERR))
+			LOG_ERR("RX Frame error");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static void ps2_npcx_ctrl_isr(const struct device *dev)
+{
+	uint8_t active_ch, mask;
+	struct ps2_reg *const inst = HAL_PS2_INSTANCE(dev);
+	struct ps2_npcx_ctrl_data *const data = DRV_DATA(dev);
+
+	/*
+	 * ACH = 1 : CHannel 0
+	 * ACH = 2 : CHannel 1
+	 * ACH = 4 : CHannel 2
+	 * ACH = 5 : CHannel 3
+	 */
+	active_ch = GET_FIELD(inst->PSTAT, NPCX_PSTAT_ACH);
+	active_ch = active_ch > 2 ? (active_ch - 2) : (active_ch - 1);
+	LOG_DBG("ACH: %d\n", active_ch);
+
+	/*
+	 * Inhibit PS/2 transaction of the other non-active channels by
+	 * pulling down the clock signal
+	 */
+	mask = ~NPCX_PSOSIG_CLK_MASK_ALL | BIT(NPCX_PSOSIG_CLK(active_ch));
+	inst->PSOSIG &= mask;
+
+	/* PS/2 Start of Transaction */
+	if (IS_BIT_SET(inst->PSTAT, NPCX_PSTAT_SOT) &&
+	    IS_BIT_SET(inst->PSIEN, NPCX_PSIEN_SOTIE)) {
+		/*
+		 * Once set, SOT is not cleared until the shift mechanism
+		 * is reset. Therefore, SOTIE should be cleared on the
+		 * first occurrence of an SOT interrupt.
+		 */
+		inst->PSIEN &= ~BIT(NPCX_PSIEN_SOTIE);
+		LOG_DBG("SOT");
+		/* PS/2 End of Transaction */
+	} else if (IS_BIT_SET(inst->PSTAT, NPCX_PSTAT_EOT)) {
+		inst->PSIEN &= ~BIT(NPCX_PSIEN_EOTIE);
+
+		/*
+		 * Clear the CLK of active channel to reset
+		 * the shift mechanism
+		 */
+		inst->PSOSIG &= ~BIT(NPCX_PSOSIG_CLK(active_ch));
+
+		/* Tx is done */
+		if (IS_BIT_SET(inst->PSCON, NPCX_PSCON_XMT)) {
+			/* Change the PS/2 module to receive mode */
+			inst->PSCON &= ~BIT(NPCX_PSCON_XMT);
+			k_sem_give(&data->tx_sync_sem);
+		} else {
+			if (ps2_npcx_ctrl_is_rx_error(dev) == 0) {
+				ps2_callback_t callback;
+				uint8_t data_in = inst->PSDAT;
+
+				LOG_DBG("Recv:0x%02x", data_in);
+				callback = data->callback_isr[active_ch];
+				if (callback != NULL)
```

Comment:
Is this allowed to be NULL ?

Reply:
The callback is assigned after the PS/2 channel is configured.
If someone enables the channel without configuring it first, the callback is NULL.